core/src/main/java/com/linecorp/armeria/common/Flags.java

```diff
@@ -520,31 +522,58 @@ public static TransportType transportType() {
      *
      * <p>This flag is enabled by default for supported platforms. Specify the
      * {@code -Dcom.linecorp.armeria.useOpenSsl=false} JVM option to disable it.
+     *
+     * @deprecated Use {@link #tlsEngineType()} and {@code -Dcom.linecorp.armeria.tlsEngineType=openssl}.
      */
+    @Deprecated
     public static boolean useOpenSsl() {
-        if (useOpenSsl != null) {
-            return useOpenSsl;
+        return tlsEngineType() == TlsEngineType.OPENSSL;
+    }
+
+    /**
+     * Returns the {@link TlsEngineType} that will be used for processing TLS connections.
+     *
+     * <p>The default value of this flag is "openssl", which means the default {@link TlsEngineType} will
+     * be used. Specify the {@code -Dcom.linecorp.armeria.tlsEngineType=<jdk|openssl>} JVM option to override
+     * the default.</p>
+     */
+    public static TlsEngineType tlsEngineType() {
+        if (TLS_ENGINE_TYPE != null) {
+            return TLS_ENGINE_TYPE;
         }
         setUseOpenSslAndDumpOpenSslInfo();
-        return useOpenSsl;
+        return TLS_ENGINE_TYPE;
     }
 
     private static void setUseOpenSslAndDumpOpenSslInfo() {
-        final boolean useOpenSsl = getValue(FlagsProvider::useOpenSsl, "useOpenSsl");
-        if (!useOpenSsl) {
+        final Boolean useOpenSsl = getUserValue(FlagsProvider::useOpenSsl, "useOpenSsl");
+        final TlsEngineType tlsEngineType = getUserValue(FlagsProvider::tlsEngineType, "tlsEngineType");
+
+        if (useOpenSsl == null) {
+            TLS_ENGINE_TYPE = tlsEngineType != null ? tlsEngineType : TlsEngineType.OPENSSL;
+        } else if (tlsEngineType == null) {
+            TLS_ENGINE_TYPE = useOpenSsl ? TlsEngineType.OPENSSL : TlsEngineType.JDK;
+        } else {
+            if (useOpenSsl != (tlsEngineType == TlsEngineType.OPENSSL)) {
+                logger.warn("useOpenSsl({}) and tlsEngineType({}) are incompatible, tlsEngineType will be used",
```

Comment:
This won't be true when `tlsEngineType` is `OPENSSL` and `OpenSsl.isAvailable()` returns `false`.
How about just warning about the incompatibility or checking the condition here?
```java
final TlsEngineType selected;
if (Boolean.FALSE.equals(useOpenSsl) || !OpenSsl.isAvailable()) {
    selected = TlsEngineType.JDK;
} else {
    selected = TlsEngineType.OPENSSL;
}
// log using selected
```

Reply:
Thank you for pointing out the missing part. I will check using the `selected` and `tlsengineType`(provided by user) and warn with log if they are different.

Reply:
@minwoox Thank you for pointing the missing part 😄 . 

What do you think about changing the sequence of validation like below(bring the `!OpenSsl.isAvailable()` to the foremost)? I think we can lower the burden of keeping `OpenSsl.isAvailable()` in mind. 
as-is 
```
        if (useOpenSsl == null) {
            tlsEngineType = tlsEngineType != null ? tlsEngineType : TlsEngineType.OPENSSL;
        } else if (tlsEngineType == null) {
            tlsEngineType = useOpenSsl ? TlsEngineType.OPENSSL : TlsEngineType.JDK;
        } else {
            if (useOpenSsl != (tlsEngineType == TlsEngineType.OPENSSL)) {
                logger.warn("useOpenSsl({}) and tlsEngineType({}) are incompatible, tlsEngineType will be used",
                            useOpenSsl, tlsEngineType);
            }
            tlsEngineType = tlsEngineType;
        }

        if (tlsEngineType != TlsEngineType.OPENSSL) {
            // OpenSSL explicitly disabled
            dumpOpenSslInfo = false;
            return;
        }
        if (!OpenSsl.isAvailable()) {
            final Throwable cause = Exceptions.peel(OpenSsl.unavailabilityCause());
            logger.info("OpenSSL not available: {}", cause.toString());
            tlsEngineType = TlsEngineType.JDK;
            dumpOpenSslInfo = false;
            return;
        }
```

to-be 
```
        if (!OpenSsl.isAvailable()) {
            final Throwable cause = Exceptions.peel(OpenSsl.unavailabilityCause());
            logger.info("OpenSSL not available: {}", cause.toString());
            tlsEngineType = TlsEngineType.JDK;
            dumpOpenSslInfo = false;
            return;
        }
        
        if (useOpenSsl == null) {
            tlsEngineType = tlsEngineType != null ? tlsEngineType : TlsEngineType.OPENSSL;
        } else if (tlsEngineType == null) {
            tlsEngineType = useOpenSsl ? TlsEngineType.OPENSSL : TlsEngineType.JDK;
        } else {
            if (useOpenSsl != (tlsEngineType == TlsEngineType.OPENSSL)) {
                logger.warn("useOpenSsl({}) and tlsEngineType({}) are incompatible, tlsEngineType will be used",
                            useOpenSsl, tlsEngineType);
            }
            tlsEngineType = tlsEngineType;
        }

        if (tlsEngineType != TlsEngineType.OPENSSL) {
            // OpenSSL explicitly disabled
            dumpOpenSslInfo = false;
            return;
        }
```

Reply:
Sorry about the late reply.
The new logic will print out the log about OpenSSL even if the user doesn't care about it.
However, I think it's acceptable and this will simplify the logic so looks okay to me. 😉 
Thanks!